test/e2e/retry.go

```diff
@@ -35,18 +36,33 @@ const (
 
 // retryWithTimeout retries a function that returns an error until a timeout is reached
 func retryWithTimeout(interval, timeout time.Duration, fn func() error) error {
+	result := make(chan error, 1)
+
+	ctx, cancelFunc := context.WithTimeout(context.Background(), timeout)
+	defer cancelFunc()
+
 	var pollError error
-	err := wait.PollImmediate(interval, timeout, func() (bool, error) {
-		pollError = nil
-		err := fn()
-		if err != nil {
-			pollError = err
-			return false, nil //nolint:nilerr // We don't want to return err here
+	go func() {
+		result <- wait.PollInfiniteWithContext(ctx, interval, func(ctx context.Context) (bool, error) {
+			pollError = nil
+			err := fn()
+			if err != nil {
+				pollError = err
+				return false, nil //nolint:nilerr // We don't want to return err here
+			}
+			return true, nil
+		})
+	}()
+
+	select {
+	case <-time.After(timeout):
```

Comment:
This was a great question, thanks for asking this. Feels like I went into a deep rabbit hole of improving `retryWithTimeout()` and I tried all the below scenarios. 

But I still need to improve this retry logic. `fn()` leaks in current implementation. TL;DR Refer https://go.dev/play/p/XP5OPsHZzoB

----


Tried a couple of with and without context. And here are my findings

- https://go.dev/play/p/XP5OPsHZzoB uses no context and relies just on timeout. 
	- notice that the fn() is running in the background even after the control has returned from `retryWithTimeout()`.
	- We must mitigate that fn() leaking away. 
	- This is current how the current PR is based off of. 

- https://go.dev/play/p/sHVgOeaSrZ3 uses context and timeout.
	- notice we run into panic when context has timed out. One way to avoid this is via avoiding closing channel result.
	- But it is not the right approach, we should mitigate it.

- https://go.dev/play/p/fV0elxeRrOO uses context and timeout too. But the condition function is also checking for context.
	- notice that we run into panic saying the channel is closed. 

- https://go.dev/play/p/SeN3mVDnBvD uses no context, and relies on timeout. 
	- however, the fn() keeps running. the mitigation efforts were on clearing returning false when encountering error from `fn()`

Reply:
One of the ideas, which seemed very promising to me, was changing `fn()` to `fn(cox context.Context)` and then perform its actions. 
Refer: https://go.dev/play/p/LF78IEnHgFu
Here, I am using both context and timeout. This works great provided we 
- dont defer closing the `result` channel. (I should explore a better way to do this)
- modify `fn()` to accept context and use it in its execution.